knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,180 @@
+		"eventing":     {},
+		"serving":      {},
+		"knativeKafka": {},
+	}
+)
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func TryStartTelemetry(c controller.Controller, mgr manager.Manager, stop chan struct{}, component string, api client.Client) error {
+	if os.Getenv(SkipTelemetryEnvVar) != "" {
+		return nil
+	}
+	log.Info("starting telemetry for:", "component", component)
+	// Initialize metrics before we start the corresponding controller.
+	// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+	InitializeAndTakeMetricsSnapshot(component, api)
+	// Start our controller in a goroutine so that we do not block.
+	go func() {
+		// Block until our controller manager is elected leader. We presume our
+		// entire process will terminate if we lose leadership, so we don't need
+		// to handle that.
+		<-mgr.Elected()
+		// Start our controller. This will block until it is stopped
+		// or the controller returns an error.
+		if err := c.Start(stop); err != nil {
+			log.Error(err, "cannot start telemetry controller for", "component", component)
+		}
+	}()
+	return nil
+}
+
+// TryStopTelemetry stops telemetry per component either Eventing, KnativeKafka or Serving
+// When called it assumes that we are reconciling a deletion event.
+func TryStopTelemetry(stop *chan struct{}, component string) {
+	if os.Getenv(SkipTelemetryEnvVar) != "" {
+		return
+	}
+	log.Info("stopping telemetry for:", "component", component)
+	// Stop the telemetry controller
+	// the lock above makes sure we close once and not panic
+	close(*stop)
+	// Can't use a closed channel
+	*stop = make(chan struct{})
+	// Remove snapshot entries for the components so that snapshot is does not get
+	// unbounded since the telemetry controller can be restarted multiple times.
+	metricsSnapshot[component].Range(func(key interface{}, value interface{}) bool {
+		metricsSnapshot[component].Delete(key)
+		return true
+	})
+}
+
+// InitializeAndTakeMetricsSnapshot is used for taking a global snapshot of metrics
+// before we start any telemetry controller. If the operator is restarted
+// client metrics will not reflect the state in the cluster so there is need to restore current state.
+// It should run each time operator is initialized and for each trial to start a Telemetry controller.
+// Snapshot is used for skipping events already counted but also received later after
+// telemetry controllers are started. Metrics are initialized here and kept around as long as the
+// Serverless Operator is running. No metric label is removed when components are removed so that
+// if any component eg. Eventing is removed and CRDs exist, it will show existing CRs correctly.
+// Telemetry should be used in the background so no error is returned here.
+func InitializeAndTakeMetricsSnapshot(component string, api client.Client) {
```

Comment:
Is this **really** necessary? I suppose we could have slightly skewed metrics for a reporting period if we don't do this (as the controller does its global resync on startup), but is that slight skew really worth the trouble? :thinking:

Reply:
If I create N objects eg 1000 and then restart the operator I get 0. This is a big difference.

Reply:
You get that 0 value until the controller has fully resynced, no?

Reply:
What if it does not happen and we get 0 for long?

Reply:
Why would it not happen?